src/Petri.jl

```diff
@@ -0,0 +1,101 @@
+module Petri
+
+using Catlab
+using Catlab.CategoricalAlgebra.CSets
+using Catlab.Present
+using Catlab.Theories
+
+# Petri Nets
+############
+
+@present TheoryPetri(FreeCategory) begin
+    T::Ob
+    S::Ob
+    I::Ob
+    O::Ob
+
+    it::Hom(I,T)
+    is::Hom(I,S)
+    ot::Hom(O,T)
+    os::Hom(O,S)
+end
+
+const AbstractPetri = AbstractCSetType(TheoryPetri)
+const Petri = CSetType(TheoryPetri, index=[:is,:it,:os,:ot])
+
+ns(p::AbstractPetri) = nparts(p,:S)
+nt(p::AbstractPetri) = nparts(p,:T)
+ni(p::AbstractPetri) = nparts(p,:I)
+no(p::AbstractPetri) = nparts(p,:O)
+
+add_species!(p::AbstractPetri) = add_part!(p,:S)
+add_species!(p::AbstractPetri,n) = add_parts!(p,:S,n)
+
+add_transition!(p::AbstractPetri) = add_part!(p,:T)
+add_transitions!(p::AbstractPetri,n) = add_parts!(p,:T,n)
+
+add_input!(p::AbstractPetri,t,s) = add_part!(p,:I,(is=s,it=t))
+add_inputs!(p::AbstractPetri,n,t,s) = add_parts!(p,:I,n,(is=s,it=t))
+
+add_output!(p::AbstractPetri,t,s) = add_part!(p,:O,(os=s,ot=t))
+add_output!(p::AbstractPetri,n,t,s) = add_parts!(p,:O,n,(os=s,ot=t))
+
+# Note: although indexing makes this pretty fast, it is often faster to bulk-convert
+# the Petri net into a transition matrix, if you are working with all of the transitions
+inputs(p::AbstractPetri,t) = subpart(p,incident(p,t,:it),:is)
+outputs(p::AbstractPetri,t) = subpart(p,incident(p,t,:ot),os)
+
+totransmat(p::AbstractPetri) = begin
+    A = zeros(Int64,(nt(p),ns(p),2))
+    for i in 1:ni(p)
+        A[subpart(p,i,:it),subpart(p,i,:is),1] += 1
+    end
+    for i in 1:no(p)
+        A[subpart(p,i,:ot),subpart(p,o,:os),2] += 1
+    end
+    A
+end
+
+fromtransmat(A::Array{Int64,3}) = begin
```

Comment:
Should this be a constructor? ex. `Petri(A::Array{T, 3}) where {T<: Integer}`

Reply:
See my previous reply.